apollo-router/src/apollo_studio_interop/mod.rs

```diff
@@ -279,6 +282,117 @@ pub(crate) fn generate_extended_references(
     generator.generate_extended_references()
 }
 
+#[tracing::instrument(skip_all, level = "debug")]
+pub(crate) fn extract_enums_from_response(
+    query: Arc<Query>,
+    operation_name: Option<&str>,
+    schema: &Valid<Schema>,
+    response_body: &Object,
+    result_set: &mut ReferencedEnums,
+) {
+    if let Some(operation) = query.operation(operation_name) {
+        extract_enums_from_selection_set(
+            &operation.selection_set,
+            &query.fragments,
+            schema,
+            response_body,
+            result_set,
+        );
+    }
+}
+
+fn add_enum_value_to_map(
+    enum_name: &Name,
+    enum_value: &JsonValue,
+    referenced_enums: &mut ReferencedEnums,
+) {
+    match enum_value {
+        JsonValue::String(val_str) => {
+            let enum_name_str = enum_name.to_string();
+            // Not using entry API here due to performance impact
+            let enum_name_stats = match referenced_enums.get_mut(&enum_name_str) {
+                Some(existing_stats) => existing_stats,
+                None => {
+                    referenced_enums.insert(enum_name_str.clone(), HashSet::new());
+                    referenced_enums
+                        .get_mut(&enum_name_str)
+                        .expect("value is expected to be in map")
+                }
+            };
+
+            enum_name_stats.insert(val_str.as_str().to_string());
+        }
+        JsonValue::Array(val_list) => {
+            for val in val_list {
+                add_enum_value_to_map(enum_name, val, referenced_enums);
+            }
+        }
+        _ => {}
+    }
+}
+
+fn extract_enums_from_selection_set(
+    selection_set: &[SpecSelection],
+    fragments: &Fragments,
+    schema: &Valid<Schema>,
+    selection_response: &Object,
+    result_set: &mut ReferencedEnums,
+) {
```

Comment:
this feels like the visitor created for demand control: https://github.com/apollographql/router/blob/dev/apollo-router/src/graphql/visitor.rs

@BrynCooke we should probably update this to move to the visitor after we merge (to make sure it won't block the release)

Reply:
I tried using a form of the visitor pattern a while ago but I can't remember why it didn't work for me. Maybe it was not available in router source before and just in federation, or maybe I'm using something that's not available from the visitor. In any case I'm happy to revisit this once all the changes are merged.